src/operator/keycloak.ts

```diff
@@ -601,178 +552,149 @@ async function internalIdp(api: KeycloakApi, connection: KeycloakConnection) {
+  api: KeycloakApi,
+  user: UserRepresentation,
+  groupsByName: Record<string, string>,
   teamGroups: string[],
 ): Promise<void> {
-  try {
-    const { body: currentKeycloakGroups } = await api.groups.realmGroupsGet(keycloakRealm)
-    const { body: existingUserGroups } = await api.users.realmUsersIdGroupsGet(keycloakRealm, existingUser.id as string)
-
-    await Promise.all(
-      teamGroups.map(async (teamGroup) => {
-        const existingGroup = existingUserGroups.find((el) => el.name === teamGroup)
-
-        if (!existingGroup) {
-          const groupId = currentKeycloakGroups.find((el) => el.name === teamGroup)?.id
-          if (groupId) {
-            await api.users.realmUsersIdGroupsGroupIdPut(keycloakRealm, existingUser.id as string, groupId as string)
-          }
+  const userId = user.id!
+  const existingUserGroups = (await api.users.realmUsersIdGroupsGet(keycloakRealm, userId)).body
+  const userGroupIds: Set<string> = new Set(existingUserGroups.map((userGroup) => groupsByName[userGroup.name]))
+  const teamGroupIds: Set<string> = new Set()
+  let groupUpdates = 0
+  await Promise.all(
+    teamGroups.map(async (teamGroup): Promise<void> => {
+      const teamGroupId = groupsByName[teamGroup]
+      if (teamGroupId) {
+        if (!userGroupIds.has(teamGroupId)) {
+          console.info(`Adding user ${user.email} to ${teamGroup}`)
+          await api.users.realmUsersIdGroupsGroupIdPut(keycloakRealm, userId, teamGroupId)
+          groupUpdates += 1
         }
-      }),
-    )
-  } catch (error) {
-    throw extractError('adding user groups', error)
+        teamGroupIds.add(teamGroupId)
+      } else {
+        console.info(`Group ${teamGroup} does not exist, skipping assignment`)
+      }
+    }),
+  )
+  await Promise.all(
+    existingUserGroups.map(async (userGroup): Promise<void> => {
+      const userGroupId = groupsByName[userGroup.name]
+      if (!teamGroupIds.has(userGroupId)) {
+        console.info(`Removing user ${user.email} from ${userGroup.name}`)
+        await api.users.realmUsersIdGroupsGroupIdDelete(keycloakRealm, userId, userGroupId)
+        groupUpdates += 1
+      }
+    }),
+  )
+  if (!groupUpdates) {
+    console.log(`No groups updated for user ${user.email}`)
   }
 }
 
-async function createUpdateUser(api: any, user: any) {
-  const { email, firstName, lastName, groups, initialPassword } = user
-  const userConf = createTeamUser(email, firstName, lastName, groups, initialPassword)
-  const existingUsersByUserEmail = (await doApiCall([], `Getting users`, () =>
-    api.users.realmUsersGet(keycloakRealm, false, `${email}`),
-  )) as UserRepresentation[]
+async function createUpdateUser(api: any, userConf: UserRepresentation): Promise<void> {
+  const groups = userConf.groups as string[]
+  const { email } = userConf
+  console.info(`Getting users for ${email}`)
+  const existingUsersByUserEmail = (await api.users.realmUsersGet(keycloakRealm, false, email))
```

Comment:
Would do it like this so its consistent with line 674. I also like this better as you see the Type directly
```suggestion
  const existingUsersByUserEmail: UserRepresentation[] = (await api.users.realmUsersGet(keycloakRealm, false, email)).body
```

Reply:
Updating this accordingly.